python/tvm/relay/backend/graph_runtime_codegen.py

```diff
@@ -129,10 +131,20 @@ def add_node(self, node, expr):
         """
         checked_type = expr.checked_type
         # setup storage ids
-        assert expr in self.storage_map
-        node.attrs["storage_id"] = [
-            x.value for x in self.storage_map[expr]
-        ]
+        assert expr in self.storage_device_map
+        storage_device_ids = self.storage_device_map[expr]
+        assert len(storage_device_ids) == 2
+        node.attrs["storage_id"] = [x.value for x in storage_device_ids[0]]
+        device_ids = [x.value for x in storage_device_ids[1]]
+        num_unknown_devices = device_ids.count(0)
+        if num_unknown_devices != 0 and num_unknown_devices != len(device_ids):
+            raise RuntimeError("The graph contains not annotated nodes for "
+                               "heterogeneous execution. All nodes must be "
+                               "annotated.")
+
+        # Add the `device_index` attribute when the graph is annotated.
```

Comment:
what's this for?

Reply:
@yzhliu This is to make sure that either all or none nodes have device_id mapping.